Source/WTF/wtf/cocoa/NSURLExtras.mm

```diff
@@ -167,38 +167,35 @@ static String decodePercentEscapes(const String& string)
     return [NSURL URLWithString:@""];
 }
 
-static NSData *dataWithUserTypedString(NSString *string)
+static RetainPtr<NSData> dataWithUserTypedString(NSString *string)
 {
-    NSData *userTypedData = [string dataUsingEncoding:NSUTF8StringEncoding];
+    RetainPtr userTypedData = [string dataUsingEncoding:NSUTF8StringEncoding];
     ASSERT(userTypedData);
     
-    const UInt8* inBytes = static_cast<const UInt8 *>([userTypedData bytes]);
-    int inLength = [userTypedData length];
-    if (!inLength)
+    auto inBytes = span(userTypedData.get());
+    if (inBytes.empty())
         return nil;
 
-    CheckedInt32 mallocLength = inLength;
+    CheckedInt32 mallocLength = inBytes.size();
     mallocLength *= 3; // large enough to %-escape every character
     if (mallocLength.hasOverflowed())
         return nil;
     
-    char* outBytes = static_cast<char *>(malloc(mallocLength));
-    char* p = outBytes;
-    int outLength = 0;
-    for (int i = 0; i < inLength; i++) {
-        UInt8 c = inBytes[i];
+    auto outBytes = MallocSpan<char, FastMalloc>::malloc(mallocLength);
```

Comment:
How important is it to not copy the string? If we don’t need that optimization then we could use `Vector` to build the string up, maybe even with inline capacity, and then make the `NSData` from the vector  and not have to write mildly tricky code.

Reply:
I switched to a Vector but still avoided the copy. I think the code is a little less tricky.